ingress/ingress.go

```diff
@@ -0,0 +1,150 @@
+	"github.com/argoproj/argo-rollouts/pkg/apis/rollouts/v1alpha1"
+	informers "github.com/argoproj/argo-rollouts/pkg/client/informers/externalversions/rollouts/v1alpha1"
+	controllerutil "github.com/argoproj/argo-rollouts/utils/controller"
+	ingressutil "github.com/argoproj/argo-rollouts/utils/ingress"
+	logutil "github.com/argoproj/argo-rollouts/utils/log"
+)
+
+const (
+	ingressIndexName = "byIngress"
+)
+
+type IngressController struct {
+	kubeclientset    kubernetes.Interface
+	rolloutsIndexer  cache.Indexer
+	rolloutSynced    cache.InformerSynced
+	ingressLister    extentionslisters.IngressLister
+	ingressSynced    cache.InformerSynced
+	rolloutWorkqueue workqueue.RateLimitingInterface
+	ingressWorkqueue workqueue.RateLimitingInterface
+	resyncPeriod     time.Duration
+
+	metricServer   *metrics.MetricsServer
+	enqueueRollout func(obj interface{})
+}
+
+// NewIngressController returns a new ingress controller
+func NewIngressController(
+	kubeclientset kubernetes.Interface,
+	ingressInformer extensionsinformers.IngressInformer,
+	rolloutsInformer informers.RolloutInformer,
+	resyncPeriod time.Duration,
+	rolloutWorkQueue workqueue.RateLimitingInterface,
+	ingressWorkQueue workqueue.RateLimitingInterface,
+	metricServer *metrics.MetricsServer) *IngressController {
+
+	controller := &IngressController{
+		kubeclientset:   kubeclientset,
+		rolloutsIndexer: rolloutsInformer.Informer().GetIndexer(),
+		rolloutSynced:   rolloutsInformer.Informer().HasSynced,
+		ingressLister:   ingressInformer.Lister(),
+		ingressSynced:   ingressInformer.Informer().HasSynced,
+
+		rolloutWorkqueue: rolloutWorkQueue,
+		ingressWorkqueue: ingressWorkQueue,
+		resyncPeriod:     resyncPeriod,
+		metricServer:     metricServer,
+	}
+
+	util.CheckErr(rolloutsInformer.Informer().AddIndexers(cache.Indexers{
+		ingressIndexName: func(obj interface{}) (strings []string, e error) {
+			if rollout, ok := obj.(*v1alpha1.Rollout); ok {
+				return ingressutil.GetRolloutIngressKeys(rollout), nil
+			}
+			return []string{}, nil
+		},
+	}))
+
+	ingressInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			controllerutil.Enqueue(obj, ingressWorkQueue)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			controllerutil.Enqueue(newObj, ingressWorkQueue)
+		},
+		DeleteFunc: func(obj interface{}) {
+			controllerutil.Enqueue(obj, ingressWorkQueue)
+		},
+	})
+	controller.enqueueRollout = func(obj interface{}) {
+		controllerutil.EnqueueRateLimited(obj, rolloutWorkQueue)
+	}
+
+	return controller
+}
+
+func (c *IngressController) Run(threadiness int, stopCh <-chan struct{}) error {
+	log.Info("Starting Ingress workers")
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(func() {
+			controllerutil.RunWorker(c.ingressWorkqueue, logutil.ServiceKey, c.syncIngress, c.metricServer)
+		}, time.Second, stopCh)
+	}
+
+	log.Info("Started Ingress workers")
+	<-stopCh
+	log.Info("Shutting down Ingress workers")
```

Comment:
where does the shutting down actually happen?  You don't seem to be doing anything here. Might be worth documenting.

Reply:
Pretty sure this was copied from the other controllers, so he was just following our lead. Shutdown would only happen when controller is being gracefully terminated.